### PR TITLE
FIR checker: differentiate UNSAFE_CALL from INAPPLICABLE_CANDIDATE

### PR DESCRIPTION
**Pull request by jsjeon**
When a receiver is nullable, which should be reported as `UNSAFE_CALL` or its variants, FIR resolution just left inapplicable candidate error. With recording expected/actual type of receiver if available, we can differentiate unsafe calls from too general inapplicable candidate errors.

This PR adds support diagnostics:
* UNSAFE_CALL
* COMPONENT_FUNCTION_ON_NULLABLE

### CODE REVIEW

**Comment by jsjeon** [comment]
I examined all (~180) test data changes, and discovered many interesting issues. Will leave comments directly. Feel free to double-check, but here is the bottom line: all those changes _were_ `INAPPLICABLE_CANDIDATE`, and replaced with either `UNSAFE_CALL` or `COMPONENT_FUNCTION_ON_NULLABLE`. Some of them need more differentiation, e.g., unsafe implicit invoke, infix, operator, iterator on nullable, etc. I noted down which one needs such improvements, and will do so gradually.

File: compiler/testData/diagnostics/tests/smartCasts/kt32358_3.fir.kt [hunk 1]
```diff
@@ -6,5 +6,5 @@ fun <R> callIt(fn: () -> R): R = TODO()
 
 fun smartIt(p1: String?, p2: String?) {
     p1 ?: callIt { TODO() }
-    p1.<!INAPPLICABLE_CANDIDATE!>length<!> // smartcast
+    p1.<!UNSAFE_CALL!>length<!> // smartcast
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/smartCasts/kt32358_3.fir.kt:9]: Filed https://youtrack.jetbrains.com/issue/KT-44510

File: compiler/testData/diagnostics/tests/smartCasts/elvis/basicOn.fir.kt [hunk 2]
```diff
@@ -11,13 +11,13 @@ interface Order {
 fun foo(o: Any) {
     val order = o as? Order
     if (order?.expired ?: false) {
-        order.<!INAPPLICABLE_CANDIDATE!>doSomething<!>()
+        order.<!UNSAFE_CALL!>doSomething<!>()
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/smartCasts/elvis/basicOn.fir.kt:14]: Filed https://youtrack.jetbrains.com/issue/KT-44511

File: compiler/testData/diagnostics/testsWithStdLib/smartcasts/letChangesToNullComplex.fir.kt [hunk 3]
```diff
@@ -5,7 +5,7 @@ fun bar(z: String?) = z
 fun foo(y: String?) {
     var x: String? = ""
     if (x != null) {
-        bar(y?.let { x = null; it }).<!INAPPLICABLE_CANDIDATE!>length<!>
+        bar(y?.let { x = null; it }).<!UNSAFE_CALL!>length<!>
         x.length // Smart cast is not possible
```
Comment (jsjeon) [inline compiler/testData/diagnostics/testsWithStdLib/smartcasts/letChangesToNullComplex.fir.kt:9]: Filed https://youtrack.jetbrains.com/issue/KT-44512

Also one below (`withChangesToNull`)

File: compiler/testData/diagnostics/tests/smartCasts/safecalls/safeAccessReceiverNotNull.fir.kt [hunk 4]
```diff
@@ -18,7 +18,7 @@ fun kt6840_2(s: String?) {
 
 fun kt1635(s: String?) {
     s?.hashCode()!!
-    s.<!INAPPLICABLE_CANDIDATE!>hashCode<!>()
+    s.<!UNSAFE_CALL!>hashCode<!>()
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/smartCasts/safecalls/safeAccessReceiverNotNull.fir.kt:21]: Filed https://youtrack.jetbrains.com/issue/KT-44513

File: compiler/testData/diagnostics/tests/smartCasts/objectLiterals/kt7110.fir.kt [hunk 5]
```diff
@@ -7,5 +7,5 @@ abstract class A(val s: String) {
 fun foo(o: String?): Int {
     val a = object : A(o!!){}
     a.bar()
-    return o.<!INAPPLICABLE_CANDIDATE!>length<!>
+    return o.<!UNSAFE_CALL!>length<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/smartCasts/objectLiterals/kt7110.fir.kt:10]: Filed https://youtrack.jetbrains.com/issue/KT-44515

File: compiler/testData/diagnostics/tests/smartCasts/comparisonUnderAnd.fir.kt [hunk 6]
```diff
@@ -6,30 +6,30 @@ fun foo(x : String?, y : String?) {
...
-        x.<!INAPPLICABLE_CANDIDATE!>length<!>
-        y.<!INAPPLICABLE_CANDIDATE!>length<!>
+        x.<!UNSAFE_CALL!>length<!>
+        y.<!UNSAFE_CALL!>length<!>
     }
     else {
         // y == null but x != y
-        x.<!INAPPLICABLE_CANDIDATE!>length<!>
-        y.<!INAPPLICABLE_CANDIDATE!>length<!>
+        x.<!UNSAFE_CALL!>length<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/smartCasts/comparisonUnderAnd.fir.kt:18]: See comment here: https://youtrack.jetbrains.com/issue/KT-41015#focus=Comments-27-4649573.0-0

File: compiler/testData/diagnostics/tests/smartCasts/publicVals/var.fir.kt [hunk 7]
```diff
@@ -4,10 +4,10 @@ public class X {
     public fun fn(): Int {
         if (x != null)
             // Smartcast is not possible for variable properties
-            return x.<!INAPPLICABLE_CANDIDATE!>length<!>
+            return x.<!UNSAFE_CALL!>length<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/smartCasts/publicVals/var.fir.kt:7]: These should be `SMARTCAST_IMPOSSIBLE` and I left a TODO there.

Will see/learn what _predictable_ nullability means at https://github.com/JetBrains/kotlin/commit/41ebfd025e6cc02cc5559b1a79d6335dd3831cb7 :) Any advices/suggestions are welcome!

File: compiler/testData/diagnostics/tests/controlStructures/ForLoopWithExtensionIteratorOnNullable.fir.kt [hunk 8]
```diff
@@ -7,9 +7,9 @@ operator fun <K> Container<K>.iterator(): Iterator<K> = null!!
 fun test() {
     val container: Container<String>? = null
     // Error
-    container.<!INAPPLICABLE_CANDIDATE!>iterator<!>()
+    container.<!UNSAFE_CALL!>iterator<!>()
     // for extension iterator, this code compiles, but should not
-    <!INAPPLICABLE_CANDIDATE!>for (s in container) {}<!>
+    <!UNSAFE_CALL!>for (s in container) {}<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/controlStructures/ForLoopWithExtensionIteratorOnNullable.fir.kt:12]: For my own record, this should be rather `ITERATOR_ON_NULLABLE`: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/controlStructures/ForLoopWithExtensionIteratorOnNullable.kt#L12

File: compiler/testData/diagnostics/tests/controlStructures/forWithNullableIterator.fir.kt [hunk 9]
```diff
@@ -8,5 +8,5 @@ class It {
 }
 
 fun test() {
-  <!INAPPLICABLE_CANDIDATE, INAPPLICABLE_CANDIDATE!>for (x in Coll()) {}<!>
+  <!UNSAFE_CALL, UNSAFE_CALL!>for (x in Coll()) {}<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/controlStructures/forWithNullableIterator.fir.kt:11]: For my own record, these should be rather `HAS_NEXT_FUNCTION_NONE_APPLICABLE` and `NEXT_NONE_APPLICABLE`: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/controlStructures/forWithNullableIterator.kt#L11

File: compiler/testData/diagnostics/tests/regressions/kt282.fir.kt [hunk 10]
```diff
@@ -14,6 +14,6 @@ fun f(): Unit {
     val i : Int? = null
     i <!NONE_APPLICABLE!>+<!> 1
     set + 1
-    1 <!INAPPLICABLE_CANDIDATE!>in<!> set
+    1 <!UNSAFE_CALL!>in<!> set
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/regressions/kt282.fir.kt:17]: For my own record, this and above should be `UNSAFE_OPERATOR_CALL`: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/regressions/kt282.kt#L17

File: compiler/testData/diagnostics/tests/nullabilityAndSmartCasts/InfixCallNullability.fir.kt [hunk 11]
```diff
@@ -12,30 +12,30 @@ fun test(x : Int?, a : A?) {
     x.<!NONE_APPLICABLE!>plus<!>(1)
     x?.plus(1)
     x <!NONE_APPLICABLE!>+<!> 1
-    <!INAPPLICABLE_CANDIDATE!>-<!>x
-    x.<!INAPPLICABLE_CANDIDATE!>unaryMinus<!>()
+    <!UNSAFE_CALL!>-<!>x
+    x.<!UNSAFE_CALL!>unaryMinus<!>()
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/nullabilityAndSmartCasts/InfixCallNullability.fir.kt:16]: For my own record, this file has many `UNSAFE_OPERATOR_CALL` and `UNSAFE_INFIX_CALL` examples!

https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/nullabilityAndSmartCasts/InfixCallNullability.kt

File: compiler/testData/diagnostics/tests/regressions/kt557.fir.kt [hunk 12]
```diff
@@ -8,5 +8,5 @@ fun Array<String>.length() : Int {
 }
 
 fun test(array : Array<String?>?) {
-    array?.sure<Array<String?>>().<!INAPPLICABLE_CANDIDATE!>length<!>()
+    array?.sure<Array<String?>>().<!UNSAFE_CALL!>length<!>()
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/regressions/kt557.fir.kt:11]: To be honest, no idea yet.

https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/regressions/kt557.kt#L11

At DOT operator, the receiver has `TYPE_MISMATCH`, and the call is reported as `UNRESOLVED_REFERENCE_WRONG_RECEIVER` (well, because type mismatches, hm). We don't have a type checker that reports `TYPE_MISMATCH`, don't we?

File: compiler/testData/diagnostics/testsWithStdLib/coroutines/restrictSuspension/outerYield_1_3.fir.kt [hunk 13]
```diff
@@ -87,7 +87,7 @@ fun test() {
                 this@with.<!UNRESOLVED_REFERENCE!>yield<!>("")
 
                 yield2("")
-                this@with.<!INAPPLICABLE_CANDIDATE!>yield2<!>("")
+                this@with.<!UNSAFE_CALL!>yield2<!>("")
```
Comment (jsjeon) [inline compiler/testData/diagnostics/testsWithStdLib/coroutines/restrictSuspension/outerYield_1_3.fir.kt:90]: Similar to `kt557`, this should be `UNRESOLVED_REFERENCE_WRONG_RECEIVER`: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/testsWithStdLib/coroutines/restrictSuspension/outerYield_1_3.kt#L90

File: compiler/testData/diagnostics/tests/extensions/throwOutCandidatesByReceiver.fir.kt [hunk 14]
```diff
@@ -8,7 +8,7 @@ fun <R> List<R>.a() {}
 
 fun test1(i: Int?) {
     1.<!INAPPLICABLE_CANDIDATE!>a<!>()
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/extensions/throwOutCandidatesByReceiver.fir.kt:10]: Just for the record, many errors in this file should be rather `UNRESOLVED_REFERENCE_WRONG_RECEIVER`: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/extensions/throwOutCandidatesByReceiver.kt#L10

File: compiler/testData/diagnostics/tests/declarationChecks/MultiDeclarationErrors.fir.kt [hunk 15]
```diff
@@ -11,7 +11,7 @@ class MyClass2 {}
 <!CONFLICTING_OVERLOADS!>fun MyClass2.component1()<!> = 1.3
 
 fun test(mc1: MyClass, mc2: MyClass2) {
-    val (<!INAPPLICABLE_CANDIDATE!>a<!>, b) = <!COMPONENT_FUNCTION_MISSING!>mc1<!>
+    val (a, b) = <!COMPONENT_FUNCTION_MISSING!>mc1<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/declarationChecks/MultiDeclarationErrors.fir.kt:14]: Perhaps, this is the only regression. In old FE, `a` has `OPERATOR_MODIFIER_REQUIRED` as well as `COMPONENT_FUNCTION_MISSING`. FIR should report unresolved error rather than inapplicable candidate. :\

https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/declarationChecks/MultiDeclarationErrors.kt#L14

File: compiler/testData/diagnostics/tests/resolve/invoke/errors/unsafeCallWithInvoke.fir.kt [hunk 16]
```diff
@@ -4,7 +4,7 @@
 operator fun String.invoke(i: Int) {}
 
 fun foo(s: String?) {
-    <!INAPPLICABLE_CANDIDATE!>s<!>(1)
+    <!UNSAFE_CALL!>s<!>(1)
 
-    <!INAPPLICABLE_CANDIDATE!>(s ?: null)(1)<!>
+    <!UNSAFE_CALL!>(s ?: null)(1)<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/resolve/invoke/errors/unsafeCallWithInvoke.fir.kt:9]: OI/NI behave differently: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/resolve/invoke/errors/unsafeCallWithInvoke.kt#L9

Assuming NI is default in FIR, I think we should report `UNSAFE_IMPLICIT_INVOKE_CALL` here?

File: compiler/testData/diagnostics/tests/collectionLiterals/basicCollectionLiterals.fir.kt [hunk 17]
```diff
@@ -14,11 +14,11 @@ fun test() {
 }
 
 fun check() {
-    [1, 2] <!INAPPLICABLE_CANDIDATE!>checkType<!> { <!INAPPLICABLE_CANDIDATE!>_<!><Array<Int>>() }
-    [""] <!INAPPLICABLE_CANDIDATE!>checkType<!> { <!INAPPLICABLE_CANDIDATE!>_<!><Array<String>>() }
+    [1, 2] <!UNSAFE_CALL!>checkType<!> { <!INAPPLICABLE_CANDIDATE!>_<!><Array<Int>>() }
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/collectionLiterals/basicCollectionLiterals.fir.kt:17]: There are many, many cases like this. Somehow `// !CHECK_TYPE` is not quite working well with FIR yet?

File: compiler/testData/diagnostics/tests/platformTypes/nullabilityWarnings/array.fir.kt [hunk 18]
```diff
@@ -21,11 +21,11 @@ fun test() {
     val platformJ = J.staticJ
 
     platformNN[0]
-    <!INAPPLICABLE_CANDIDATE!>platformN[0]<!>
+    <!UNSAFE_CALL!>platformN[0]<!>
     platformJ[0]
 
     platformNN[0] = 1
-    <!INAPPLICABLE_CANDIDATE!>platformN[0]  = 1<!>
+    <!UNSAFE_CALL!>platformN[0]  = 1<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/platformTypes/nullabilityWarnings/array.fir.kt:28]: For the record. In addition to `.`, this is another place where the diagnostic positioning matters: it should be array access.

**Comment by jsjeon** [comment]
https://teamcity.jetbrains.com/buildConfiguration/Kotlin_KotlinForGoogle_AggregateBranch/3283456?buildTab=tests&status=failed

It seems like `FirDiagnosticTestSpecGenerated` is not triggered by `:compiler:tests-spec:test` anymore (maybe in the middle of JUnit5 migration?), sigh... I need to examine 70 more test files; converting this PR draft while doing so.

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/neg/8.fir.kt [hunk 19]
```diff
@@ -18,7 +18,7 @@ fun case_2(x: Class?) {
     if (x?.fun_4()?.prop_8 != null) {
         <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>
         <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.fun_4()
-        <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.fun_4().<!INAPPLICABLE_CANDIDATE!>prop_8<!>
+        <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.fun_4().<!UNSAFE_CALL!>prop_8<!>
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/notLinked/dfa/neg/8.fir.kt:21]: For the record, this should be `SMARTCAST_IMPOSSIBLE` instead.

File: compiler/tests-spec/testData/diagnostics/notLinked/contracts/analysis/smartcasts/neg/6.fir.kt [hunk 20]
```diff
@@ -111,7 +111,7 @@ import contracts.*
 fun case_1(value_1: Any?, value_2: Int?) {
     value_1.case_1(value_2)
     <!AMBIGUITY!>println<!>(value_1.<!UNRESOLVED_REFERENCE!>length<!>)
-    println(value_2.<!INAPPLICABLE_CANDIDATE!>inv<!>())
+    println(value_2.<!UNSAFE_CALL!>inv<!>())
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/notLinked/contracts/analysis/smartcasts/neg/6.fir.kt:114]: The contract for the extension function `case_1` above is that, if returned, the given argument is not null. Thus, `value_2` should be not null here.

Many other unsafe calls in this file are originated from such incorrect contract propagations.

File: compiler/tests-spec/testData/diagnostics/linked/statements/assignments/p-2/neg/1.2.fir.kt [hunk 21]
```diff
@@ -8,7 +8,7 @@
  */
 fun case1() {
     val x : Case1? = Case1()
-    x.<!INAPPLICABLE_CANDIDATE!>x<!> = "0"
+    x.x = "0"
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/linked/statements/assignments/p-2/neg/1.2.fir.kt:11]: Note that this test intentionally adds `-UNSAFE_CALL` to focus on `val` reassignments.

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/52.fir.kt [hunk 22]
```diff
@@ -89,7 +89,7 @@ fun <T : String?> T.case_9() = this
 fun case_9() {
     var x: Int? = 10
     x = null
-    <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Int? & kotlin.Int?")!>x<!>.<!INAPPLICABLE_CANDIDATE!>case_9<!>()
+    <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Int? & kotlin.Int?")!>x<!>.<!UNSAFE_CALL!>case_9<!>()
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/52.fir.kt:92]: After assigning `null`, it's expected to be resolved to:
```
fun <T: String?> T.case_9() = ...
```
above, but I'd say this looks a lot like https://youtrack.jetbrains.com/issue/KT-41017

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/61.fir.kt [hunk 23]
```diff
@@ -158,7 +158,7 @@ class Case11 {
         if (y == case_12)
             throw Exception()
         <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.String?")!>y<!>
-        <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.String?")!>y<!>.<!INAPPLICABLE_CANDIDATE!>length<!>
+        <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.String?")!>y<!>.<!UNSAFE_CALL!>length<!>
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/61.fir.kt:161]: For the record, this is also `SMARTCAST_IMPOSSIBLE`.

File: compiler/tests-spec/testData/diagnostics/linked/overload-resolution/building-the-overload-candidate-set-ocs/call-with-named-parameters/p-1/pos/2.1.fir.kt [hunk 24]
```diff
@@ -63,7 +63,7 @@ fun case3() {
  */
 fun case4(marker : Marker?) {
     marker?.<!DEBUG_INFO_CALL("fqName: libPackage.Marker.foo; typeCall: function")!>foo(y=1)<!>
-    marker?.<!DEBUG_INFO_CALL("fqName: libPackage.Marker.invoke; typeCall: variable&invoke")!>foo(x=1)<!><!INAPPLICABLE_CANDIDATE!>><!><!SYNTAX!><!>
+    marker?.<!DEBUG_INFO_CALL("fqName: libPackage.Marker.invoke; typeCall: variable&invoke")!>foo(x=1)<!><!UNSAFE_CALL!>><!><!SYNTAX!><!>
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/linked/overload-resolution/building-the-overload-candidate-set-ocs/call-with-named-parameters/p-1/pos/2.1.fir.kt:66]: This one and one more below are rather resolution errors. It should be `tests.foo`, instead of these `invoke`.

The original issue was: https://youtrack.jetbrains.com/issue/KT-37391

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/25.fir.kt [hunk 25]
```diff
@@ -52,8 +52,8 @@ inline fun <reified T : CharSequence>case_4(x: Any?) {
     (x as? T)!!
     if (x is T?) {
         <!DEBUG_INFO_EXPRESSION_TYPE("T? & kotlin.Any?")!>x<!>
-        <!DEBUG_INFO_EXPRESSION_TYPE("T? & kotlin.Any?")!>x<!>.<!INAPPLICABLE_CANDIDATE!>length<!>
-        <!DEBUG_INFO_EXPRESSION_TYPE("T? & kotlin.Any?")!>x<!>.<!INAPPLICABLE_CANDIDATE!>get<!>(0)
+        <!DEBUG_INFO_EXPRESSION_TYPE("T? & kotlin.Any?")!>x<!>.<!UNSAFE_CALL!>length<!>
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/25.fir.kt:55]: This (25) and below two (26, 27) share the same issue. Filed https://youtrack.jetbrains.com/issue/KT-44559

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/30.fir.kt [hunk 26]
```diff
@@ -11,9 +11,9 @@ fun case_1(x: Class?) {
     if (x!!.prop_8?.prop_8?.prop_8?.prop_8 != null) {
         <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>
         <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.prop_8
-        <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.prop_8.<!INAPPLICABLE_CANDIDATE!>prop_8<!>
-        <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.prop_8.<!INAPPLICABLE_CANDIDATE!>prop_8<!>.<!INAPPLICABLE_CANDIDATE!>prop_8<!>
-        <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.prop_8.<!INAPPLICABLE_CANDIDATE!>prop_8<!>.<!INAPPLICABLE_CANDIDATE!>prop_8<!>.<!INAPPLICABLE_CANDIDATE!>prop_8<!>
+        <!DEBUG_INFO_EXPRESSION_TYPE("Class & Class?")!>x<!>.prop_8.<!UNSAFE_CALL!>prop_8<!>
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/30.fir.kt:14]: This (30) and below (31) share the same issue. Filed https://youtrack.jetbrains.com/issue/KT-44560

File: compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/6.fir.kt [hunk 27]
```diff
@@ -19,13 +19,13 @@ fun case_1(x: Any?) {
     val y = null
     if (x != y) {
         <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Any?")!>x<!>
-        <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Any?")!>x<!>.<!INAPPLICABLE_CANDIDATE!>equals<!>(null)
+        <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.Any?")!>x<!>.<!UNSAFE_CALL!>equals<!>(null)
```
Comment (jsjeon) [inline compiler/tests-spec/testData/diagnostics/notLinked/dfa/pos/6.fir.kt:22]: Filed https://youtrack.jetbrains.com/issue/KT-44561

There are many more variants in this file as well as: `pos/2.fir.kt` and `pos/7.fir.kt`.

**Comment by jsjeon** [comment]
> I need to examine 70 more test files

I did! and found a couple more interesting cases (also filed). In general, there are no strange `UNSAFE_CALL` additions.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt [hunk 28]
```diff
@@ -411,6 +418,18 @@ class FirDefaultErrorMessages : DefaultErrorMessages.Extension {
...
             map.put(FirErrors.WRONG_IMPLIES_CONDITION, "Wrong implies condition")
 
+            // Nullability
+            map.put(
+                UNSAFE_CALL,
+                "Only safe (?.) or non-null asserted (!!.) calls are allowed on a nullable receiver of type {0}",
+                RENDER_TYPE
+            )
+            //map.put(
+            //    UNSAFE_IMPLICIT_INVOKE_CALL,
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:428]: Forgot to mention this. I've tried this by adding a naive check that just checks an `invoke` function call, but then found several cases that require overrides check in general. There are a couple utils that are used at resolution and `fir2ir` regarding `invoke` lookup, and I guess it will require more refactoring, which I guess would be better to be a separate PR/commit.

**Comment by mglukhikh** [comment]
Merged manually. I fixed the following things:

- `UNSAFE_CALL` is no more reported for a case with `T` receiver at declaration-site
- correct positioning (at dot) added

**Comment by jsjeon** [comment]
> Merged manually. I fixed the following things:

Thank you!

> * `UNSAFE_CALL` is no more reported for a case with `T` receiver at declaration-site

https://github.com/JetBrains/kotlin/commit/0ee4f1f3934a415570dd410897e8491b231cdade

Ah, missed type parameter.

> * correct positioning (at dot) added

https://github.com/JetBrains/kotlin/commit/7d4eaefd368545c9770e293b975f2a6aab88caf1

Yay~ Thanks again!

File: compiler/testData/diagnostics/tests/regressions/kt282.fir.kt (hunk 10 above)
Comment (jsjeon) [reply compiler/testData/diagnostics/tests/regressions/kt282.fir.kt:17]: Done at https://github.com/JetBrains/kotlin/pull/4074/files#diff-30a21c37fc436b587141fc7c83bd1e5a73d90efdeac354179605d7a984bda37b

File: compiler/testData/diagnostics/tests/nullabilityAndSmartCasts/InfixCallNullability.fir.kt (hunk 11 above)
Comment (jsjeon) [reply compiler/testData/diagnostics/tests/nullabilityAndSmartCasts/InfixCallNullability.fir.kt:16]: Done at https://github.com/JetBrains/kotlin/pull/4074/files#diff-822ac7b0f078d8d87df55b187f13c41140247870c47afbf65d7464a938bbfdf6
